### ISSUE TITLE
T Prefix for ClassLib Types

### ISSUE THREAD

**Issue by WolfieWerewolf** [konsoletyper/teavm#291]
Hi Alexey,

I'm a big fan of your work, thanks for sharing.  I've resisted the urge to ask questions until I had read every line of code in your project; more than once.

Now I have a question.  The emulation library types are all prefixed with "T", presumably to avoid name conflicts with the JDK.  Sometimes, however, you consume your own "T" implementation and other times you use the JDK Type.  Object seems to be a recurring case as well as some of the exceptions.  In some cases you also have your own implementation of the exception in question.  Is there a rule for when to consume "T" Types vs JDK types?  I'm interested in extending the ClassLib and wanted to follow the correct pattern.
Many thanks in advance for any advice you can offer.  Again, I am a huge fan of your work and very much appreciate you sharing this with the community.

Kind Regards
Wolfie

**Comment by konsoletyper** [comment]
When I just started TeaVM, I had only poor support of Java stdlib. Java stdlib classes have many interdependencies between each other, i.e. `Integer.toString()` can be implemented via `StringBuilder.append(int)`. While I was implementing stdlib classes, I used `T` prefix to prevent myself from using unimplemented API. I had a convention: every time I want to use class from `java.*` package, I rather use `org.teavm.classlib.java.*.T` class. That was quite invonvenient though, since some Java constructs rely on concrete classes (`for` requires `java.lang.Iterable`, `+` requires `java.lang.String`, etc). So, when classlib implementation became mature enough, I just started using normal `java` package and just relied on tests. Note that now I have relatively good test coverage, as well as test infrastructure and good error reporting. I had none of that when I just started, so this `T` prefix was the only thing that could save me.
